src/design/plone/theme/browser/viewlets.py

```diff
@@ -161,25 +170,55 @@ def update(self):
         super(HeaderBannerViewlet, self).update()
 
         self.header_link_label = self.get_value_from_registry(
-            'header_link_label')
+            'header_link_label'
+        )
         self.header_link_url = self.get_value_from_registry('header_link_url')
         self.header_second_link_label = self.get_value_from_registry(
-            'header_second_link_label')
+            'header_second_link_label'
+        )
         self.header_second_link_url = self.get_value_from_registry(
-            'header_second_link_url')
+            'header_second_link_url'
+        )
         self.login_button_visible = self.get_bool_value_from_registry(
-            'login_button_visible')
+            'login_button_visible'
+        )
 
     def showLanguageSelector(self):
-        return (self.header_link_label and self.header_link_url) or (
-            self.header_second_link_url and self.header_second_link_url
-        ) or self.login_button_visible or self.available()
+        return (
+            (self.header_link_label and self.header_link_url)
+            or (self.header_second_link_url and self.header_second_link_url)
+            or self.login_button_visible
+            or self.available()
+        )
 
     def getUserInfos(self):
         if api.user.is_anonymous():
             return {}
         user = api.user.get_current()
         return {
             'id': user.getProperty('id'),
-            'fullname': user.getProperty('fullname') or user.getProperty('id')
+            'fullname': user.getProperty('fullname') or user.getProperty('id'),
         }
+
+
+class ContentRelatedItems(BaseContentRelatedItems):
+    def related2brains(self, related):
+        """Return a list of brains based on a list of relations. Will filter
+        relations if the user has no permission to access the content.
+
+        :param related: related items
+        :type related: list of relations
+        :return: list of catalog brains
+        """
+        catalog = api.portal.get_tool(name='portal_catalog')
+        brains = []
+        for r in related:
+            path = getattr(r, 'to_path', None)
+            if path is None:
+                # Item was deleted.  The related item should have been cleaned
+                # up, but apparently this does not happen.
+                continue
+            # the query will return an empty list if the user
+            # has no permission to see the target object
+            brains.extend(catalog(path=dict(query=path, depth=0)))
```

Comment:
Direi che puoi migliorare le performance facendo una sola chiamata al catalog:
```
paths = list(filter(None, (getattr(r, 'to_path', None) for r in related)))
if not paths:
    return []
return catalog(path=paths)
```

Reply:
@ale-rt domanda: 

> list(filter(None, (getattr(r, 'to_path', None) for r in related)))

non è più leggibile con una list-comprehension?

Reply:
Certo, anche cosi' andrebbe bene:
```
[path for path in (getattr(r, 'to_path', None) for r in related) if path]
```
ma `filter` l'hanno messo apposta ;)

Aggiungo che lo stesso risultato lo puoi ottenere in mille modi piu' leggibili, e.g. spezzando la riga in piu' righe, indentando, usando un ciclo for e poi usare `paths.append`...
Come lo scrivi non cambia il tempo di esecuzione (di solito il numero di relazioni e' basso).

Quello che fa la differenza e' diminuire il numero di chiamate al catalogo.